### ISSUE TITLE
Avoiding name conflicts within a package

### ISSUE THREAD

**Issue by lhstrh** [icyphy/lingua-franca#194]
With the new import mechanism that we're working on, name conflicts between reactor classes across files can be avoided by putting them in a different package. We need to report duplicate names among reactor classes within the same package as errors, but there are situations in which duplicates could co-exist without problem, namely if they are only used locally, within the file of their definition. Our suite of regression tests is an example of such situation.

It might, therefore, make sense to limit which reactor classes are importable by other. We could achieve this by requiring an explicit `export` modifier before any class definition that can be referenced/imported by in another file. This would also make it possible to export a composite without exporting the class definitions of the reactors it contains, for instance.

Is this something we'd want?

**Comment by edwardalee** [comment]
This sounds like a good step to me.  Being able to export a composite without exporting the components is a clear win.  It also removes the oddness that import does not import main reactors.

**Comment by cmnrd** [comment]
Sounds also good to me. I would prefer `private` and `public` modifiers rather than export though

**Comment by lhstrh** [comment]
Just to clarify, do we want to be able to export main reactors? I'm assuming that if we do, we'd simply treat them as non-main once imported. Correct?

If we do this, then perhaps `export` really is more descriptive than `public` because it's not merely a visibility modifier; it actually changes the meaning of the definition of a main reactor.

Finally, what would it mean to import a federated reactor?

**Comment by edwardalee** [comment]
I doubt exporting main reactors will prove very useful. They won't have inputs or outputs, for one thing.

**Comment by lhstrh** [comment]
I agree. I just understood your comment to imply that the reason to not import a reactor should not be because it's `main` but because it isn't exported. This left open the possibility to export main reactors. I'm also fine disallow that. So is the suggested behavior to let `export main reactor X { ...}` yield a validation error?

**Comment by edwardalee** [comment]
Yes, I would think so.

**Comment by edwardalee** [comment]
Or just don't include it in the grammar.

**Comment by lhstrh** [comment]
Better to do it during validation because it allows for a more informative error message.
